### PR TITLE
 add CRUD

### PR DESCRIPTION
**Pull request by fqliao**

### CODE REVIEW

File: libprecompiled/CRUDPrecompiled.cpp [hunk 1]
```diff
@@ -77,3 +168,84 @@ bytes CRUDPrecompiled::call(ExecutiveContext::Ptr context, bytesConstRef param,
     }
     return out;
 }
+
+void CRUDPrecompiled::parseCondition(
+    const std::string& conditionStr, const Table::Ptr& table, Condition::Ptr& condition)
```
Comment (bxq2011hust) [inline libprecompiled/CRUDPrecompiled.cpp]: `const Table::Ptr& table` this param is uncessary.

File: libprecompiled/CRUDPrecompiled.cpp [hunk 2]
```diff
@@ -77,3 +168,84 @@ bytes CRUDPrecompiled::call(ExecutiveContext::Ptr context, bytesConstRef param,
...
+                else if (*it == "limit")
+                {
+                    std::string offsetCount = OPJson[*it].asString();
+                    std::vector<std::string> offsetCountList;
+                    boost::split(offsetCountList, offsetCount, boost::is_any_of(","));
+                    size_t offset = boost::lexical_cast<size_t>(offsetCountList[0]);
+                    size_t count = boost::lexical_cast<size_t>(offsetCountList[1]);
+                    condition->limit(offset, count);
+                }
+                else
```
Comment (bxq2011hust) [inline libprecompiled/CRUDPrecompiled.cpp]: use `else if` instead `else`, use `else` to handle unexpected input.

File: libprecompiled/CRUDPrecompiled.cpp [hunk 3]
```diff
@@ -77,3 +168,84 @@ bytes CRUDPrecompiled::call(ExecutiveContext::Ptr context, bytesConstRef param,
...
+                {
+                    condition->LE(*iter, OPJson[*it].asString());
+                }
+            }
+        }
+    }
+}
+
+void CRUDPrecompiled::parseEntry(
+    const std::string& entryStr, const Table::Ptr& table, Entry::Ptr& entry)
```
Comment (bxq2011hust) [inline libprecompiled/CRUDPrecompiled.cpp]: const Table::Ptr& table this param is uncessary.

File: libprecompiled/CRUDPrecompiled.cpp (hunk 1 above)
Comment (fqliao) [reply libprecompiled/CRUDPrecompiled.cpp]: removed it

File: libprecompiled/CRUDPrecompiled.cpp (hunk 2 above)
Comment (fqliao) [reply libprecompiled/CRUDPrecompiled.cpp]: fix it

File: libprecompiled/CRUDPrecompiled.cpp (hunk 3 above)
Comment (fqliao) [reply libprecompiled/CRUDPrecompiled.cpp]: removed it

File: libstorage/MemoryTable.h [hunk 4]
```diff
@@ -408,7 +408,7 @@ class MemoryTable : public Table
     {
         std::vector<size_t> indexes;
         indexes.reserve(entries->size());
-        if (condition->getConditions()->empty())
+        if (condition->getConditions()->empty() && !condition->isDirty())
```
Comment (bxq2011hust) [inline libstorage/MemoryTable.h]: Can we do this in another way？

File: libstorage/Table.cpp [hunk 5]
```diff
@@ -231,6 +231,22 @@ void Condition::limit(size_t offset, size_t count)
 {
     m_offset = offset;
     m_count = count;
+    m_dirty = true;
+}
+
+size_t Condition::getOffset()
```
Comment (bxq2011hust) [inline libstorage/Table.cpp]: add  `getOffset()/getCount()/isDirty()` is necessary?

File: libprecompiled/CRUDPrecompiled.cpp [hunk 6]
```diff
@@ -77,3 +170,86 @@ bytes CRUDPrecompiled::call(ExecutiveContext::Ptr context, bytesConstRef param,
...
+}
+
+void CRUDPrecompiled::parseEntry(const std::string& entryStr, Entry::Ptr& entry)
+{
+    Json::Value entryJson;
+    Json::Reader reader;
+    if (!reader.parse(entryStr, entryJson))
+    {
+        PRECOMPILED_LOG(ERROR) << LOG_BADGE("CRUDPrecompiled") << LOG_DESC("entry json parse error")
+                               << LOG_KV("entry", entryStr);
```
[the remark this answers is not in the document]
Comment (fqliao) [reply libprecompiled/CRUDPrecompiled.cpp]: I add error code to notify the client.

File: libprecompiled/CRUDPrecompiled.cpp [hunk 7]
```diff
@@ -56,18 +64,103 @@ bytes CRUDPrecompiled::call(ExecutiveContext::Ptr context, bytesConstRef param,
...
+        Table::Ptr table = openTable(context, tableName);
+        if (table)
+        {
+            Entry::Ptr entry = table->newEntry();
+            parseEntry(entryStr, entry);
+            Condition::Ptr condition = table->newCondition();
+            parseCondition(conditionStr, condition);
+            int result =
+                table->update(key, entry, condition, std::make_shared<AccessOptions>(origin));
+            out = abi.abiIn("", result);
```
[the remark this answers is not in the document]
Comment (fqliao) [reply libprecompiled/CRUDPrecompiled.cpp]: Ok

File: libstorage/MemoryTable.h (hunk 4 above)
Comment (fqliao) [reply libstorage/MemoryTable.h]: I move the limit logic to CRUDPrecompiled

File: libstorage/Table.cpp (hunk 5 above)
Comment (fqliao) [reply libstorage/Table.cpp]: remove it.

**Pull request merged**